### PR TITLE
Category

### PR DESCRIPTION
**Pull request by VVMaxVV**
Added category list, where user in future can make a choice of products of which category he is interested in.

### CODE REVIEW

File: app/src/main/java/com/example/internetshop/data/repository/CategoryRepositoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.example.internetshop.domain.data.model.Category
+import com.example.internetshop.domain.data.repository.CategoryRepository
+import io.reactivex.Single
+import javax.inject.Inject
+
+class CategoryRepositoryImpl @Inject constructor(
+    private val categoryApi: CategoryApi,
+    private val categoryResponseMapper: CategoryResponseMapper
+) :
+    CategoryRepository {
```
Comment (darthoo) [inline app/src/main/java/com/example/internetshop/data/repository/CategoryRepositoryImpl.kt:15]: Looks like formatting off here. Please format the file

**Review by darthoo — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/internetshop/data/repository/CategoryRepositoryImpl.kt (hunk 1 above)
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/data/repository/CategoryRepositoryImpl.kt:15]: Fixed

File: app/src/main/java/com/example/internetshop/data/response/CategoryResponse.kt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package com.example.internetshop.data.response
+
+data class CategoryResponse(var categoryName: List<String>)
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/data/response/CategoryResponse.kt:3]: Fixed

File: app/src/main/java/com/example/internetshop/data/response/mapper/CategoryResponseMapper.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                )
+                "men's clothing" -> Category(
+                    nameOfCategory,
+                    "https://www.farmers.co.nz/INTERSHOP/static/WFS/Farmers-Shop-Site/-/Farmers-Shop/en_NZ/2021/September/FTC4030-New-Season-Cat-Tiles-7-Oct/Mens-Cat-Tiles/02-Mens-Clothing/03-Mens-Clothing-Coats-Jackets.jpg"
+                )
+                "women's clothing" -> Category(
+                    nameOfCategory,
+                    "https://media.4rgos.it/i/Argos/4221-m007-25-01-women-sweatcoat?maxW=768&qlt=75&fmt.jpeg.interlaced=true"
+                )
+                else -> Category("", "")
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/data/response/mapper/CategoryResponseMapper.kt:28]: Created and fixed

File: app/src/main/java/com/example/internetshop/data/response/mapper/CategoryResponseMapper.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package com.example.internetshop.data.response.mapper
+
+import com.example.internetshop.data.response.CategoryResponse
+import com.example.internetshop.domain.data.model.Category
+import javax.inject.Inject
+
+class CategoryResponseMapper @Inject constructor(){
+    fun toCategory(categoryResponseList: CategoryResponse): List<Category> {
+        return categoryResponseList.categoryName.map {
+            val nameOfCategory = it.capitalize()
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/data/response/mapper/CategoryResponseMapper.kt:10]: Fixed

File: app/src/main/java/com/example/internetshop/domain/data/model/Category.kt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+package com.example.internetshop.domain.data.model
+
+data class Category(val category: String, val imageUri: String)
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/domain/data/model/Category.kt:3]: Fixed

File: app/src/main/java/com/example/internetshop/domain/data/usecase/GetAuthUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.example.internetshop.domain.data.usecase
+
+import com.example.internetshop.domain.data.model.Token
+import com.example.internetshop.domain.data.model.UserCredentials
+import io.reactivex.Single
+
+interface GetAuthUseCase {
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/domain/data/usecase/GetAuthUseCase.kt:7]: Fixed

File: app/src/main/java/com/example/internetshop/domain/data/usecase/GetCategoryUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.example.internetshop.domain.data.usecase
+
+import com.example.internetshop.domain.data.model.Category
+import io.reactivex.Single
+
+interface GetCategoryUseCase {
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/domain/data/usecase/GetCategoryUseCase.kt:6]: Fixed

File: app/src/main/java/com/example/internetshop/domain/data/usecase/impl/GetCategoryUseCaseImpl.kt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.example.internetshop.domain.data.usecase.impl
+
+import com.example.internetshop.domain.data.model.Category
+import com.example.internetshop.domain.data.repository.CategoryRepository
+import com.example.internetshop.domain.data.usecase.GetCategoryUseCase
+import io.reactivex.Single
+import javax.inject.Inject
+
+class GetCategoryUseCaseImpl @Inject constructor(private val categoryRepository: CategoryRepository) :
+    GetCategoryUseCase {
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/domain/data/usecase/impl/GetCategoryUseCaseImpl.kt:10]: Fixed, but need edit config my code editor

File: app/src/main/java/com/example/internetshop/domain/di/module/AuthUseCaseModule.kt [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import com.example.internetshop.domain.data.usecase.GetAuthUseCase
+import com.example.internetshop.domain.data.usecase.impl.GetAuthUseCaseImpl
+import dagger.Module
+import dagger.Provides
+
+@Module
+class AuthUseCaseModule {
+    @Provides
+    fun getAuthUseCase(authRepository: AuthRepository): GetAuthUseCase {
+        return GetAuthUseCaseImpl(authRepository)
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/domain/di/module/AuthUseCaseModule.kt:13]: Fixed

File: app/src/main/java/com/example/internetshop/domain/di/module/CategoryUseCaseModule.kt [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.example.internetshop.domain.data.repository.CategoryRepository
+import com.example.internetshop.domain.data.usecase.GetCategoryUseCase
+import com.example.internetshop.domain.data.usecase.impl.GetCategoryUseCaseImpl
+import dagger.Module
+import dagger.Provides
+
+@Module
+class CategoryUseCaseModule {
+    @Provides
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/domain/di/module/CategoryUseCaseModule.kt:11]: Fixed

File: app/src/main/java/com/example/internetshop/model/data/di/module/AuthModule.kt [hunk 11]
```diff
@@ -27,7 +27,7 @@ class AuthModule {
     }
 
     @Provides
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/model/data/di/module/AuthModule.kt:31]: Fixed

File: app/src/main/java/com/example/internetshop/model/data/di/module/CategoryModule.kt [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    @Provides
+    fun getCategoryRepository(
+        categoryApi: CategoryApi,
+        categoryResponseMapper: CategoryResponseMapper
+    ): CategoryRepository {
+        return CategoryRepositoryImpl(categoryApi, categoryResponseMapper)
+    }
+
+    @Provides
+    fun getCategoryResponseMapper(): CategoryResponseMapper {
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/model/data/di/module/CategoryModule.kt:27]: fixed

File: app/src/main/java/com/example/internetshop/model/data/di/module/CategoryModule.kt [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+import retrofit2.Retrofit
+
+@Module
+class CategoryModule {
+    @Provides
+    fun getCategoryApi(retrofit: Retrofit): CategoryApi {
+        return retrofit.create(CategoryApi::class.java)
+    }
+
+    @Provides
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/model/data/di/module/CategoryModule.kt:18]: fixed

File: app/src/main/java/com/example/internetshop/presentation/activity/fragments/BaseDatabindingFragment.kt [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+import androidx.lifecycle.ViewModel
+import com.example.internetshop.presentation.viewModel.MultiViewModuleFactory
+import javax.inject.Inject
+
+abstract class BaseDatabindingFragment<V:ViewModel>(private val aClass: Class<V>) :Fragment() {
+
+    @Inject
+    protected lateinit var factory: MultiViewModuleFactory
+
+//    protected val viewModel by lazy {
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/activity/fragments/BaseDatabindingFragment.kt:13]: fixed

File: app/src/main/java/com/example/internetshop/presentation/activity/fragments/CategoriesFragment.kt [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+        viewModel.navEventLiveData.observe(requireActivity(), {
+            when (it) {
+                is CategoriesViewModel.CategoryEvent.OpenCategoryProductListEvent -> openCategory()
+                is CategoriesViewModel.CategoryEvent.ToastCategoryEvent -> showToast(it.text)
+            }
+        })
+
+        viewModel.categoriesLiveData.observe(requireActivity(), {
+            adapter.categoryList.add(it)
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/activity/fragments/CategoriesFragment.kt:56]: Fixed

File: app/src/main/java/com/example/internetshop/presentation/adapters/CategoryAdapter.kt [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
+        holder.bind(categoryList[position])
+        holder.itemView.setOnClickListener {
+            clickListener(categoryList[position])
+        }
+    }
+
+    override fun getItemCount(): Int {
+        return categoryList.size
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/adapters/CategoryAdapter.kt:30]: Снимок экрана 2022-01-24 в 00 45 16

File: app/src/main/java/com/example/internetshop/presentation/adapters/VerticalSpaceItemDecoration.kt [hunk 17]
```diff
@@ -0,0 +1,15 @@
+import android.graphics.Rect
+import android.view.View
+import androidx.recyclerview.widget.RecyclerView
+import androidx.recyclerview.widget.RecyclerView.ItemDecoration
+
+
+class VerticalSpaceItemDecoration(private val verticalSpaceHeight: Int) :
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/adapters/VerticalSpaceItemDecoration.kt:7]: fixed

File: app/src/main/java/com/example/internetshop/presentation/viewModel/AuthenticationViewModel.kt [hunk 18]
```diff
@@ -1,62 +1,58 @@
...
 import com.example.internetshop.presentation.utils.SingleLiveEvent
 import io.reactivex.android.schedulers.AndroidSchedulers
-import io.reactivex.functions.Consumer
 import io.reactivex.schedulers.Schedulers
 import javax.inject.Inject
 
 class AuthenticationViewModel @Inject constructor(
-    private val loginRepository: LoginRepository,
-    private val tokenPreference: TokenPreference
+    private val authRepository: AuthRepository,
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/viewModel/AuthenticationViewModel.kt]: fixed

File: app/src/main/java/com/example/internetshop/presentation/viewModel/AuthenticationViewModel.kt [hunk 19]
```diff
@@ -1,62 +1,58 @@
...
+    private val tokenPreference: TokenPreference,
+    private val getAuthUseCaseImpl: GetAuthUseCaseImpl
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/viewModel/AuthenticationViewModel.kt]: fixed

File: app/src/main/java/com/example/internetshop/presentation/viewModel/CategoriesViewModel.kt [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+import com.example.internetshop.presentation.utils.SingleLiveEvent
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.schedulers.Schedulers
+import javax.inject.Inject
+
+class CategoriesViewModel @Inject constructor(private val categoryUseCase: GetCategoryUseCase) :
+    BaseViewModel() {
+
+    val navEventLiveData = SingleLiveEvent<CategoryEvent>()
+    val categoriesLiveData = MutableLiveData<Category>()
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/viewModel/CategoriesViewModel.kt:16]: fixed

File: app/src/main/java/com/example/internetshop/presentation/viewModel/CategoriesViewModel.kt [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+    fun getCategory() {
+        compositeDisposable.add(
+            categoryUseCase.execute()
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({
+                    it.map {
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/java/com/example/internetshop/presentation/viewModel/CategoriesViewModel.kt:24]: fixed

File: app/src/main/res/layout/item_category.xml [hunk 22]
```diff
@@ -1,25 +1,51 @@
...
+            type="com.example.internetshop.domain.data.model.Category" />
+    </data>
+
+
+    <androidx.cardview.widget.CardView xmlns:android="http://schemas.android.com/apk/res/android"
+        xmlns:tools="http://schemas.android.com/tools"
+        android:layout_width="match_parent"
+        android:layout_height="@dimen/dimen_item_category_height"
+        android:background="@android:color/transparent"
+        app:cardElevation="8dp"
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/res/layout/item_category.xml]: fixed

File: app/src/main/res/layout/item_category.xml [hunk 23]
```diff
@@ -1,25 +1,51 @@
...
+        app:cardCornerRadius="8dp">
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/res/layout/item_category.xml]: fixed

File: app/src/main/res/layout/item_category.xml [hunk 24]
```diff
@@ -1,25 +1,51 @@
...
+
+        <androidx.appcompat.widget.LinearLayoutCompat
+            android:layout_width="match_parent"
+            android:layout_height="@dimen/dimen_item_category_height"
+            android:background="@color/white"
+            android:orientation="horizontal"
+            android:weightSum="2">
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/res/layout/item_category.xml:32]: fixed

File: app/src/main/res/layout/item_category.xml [hunk 25]
```diff
@@ -1,25 +1,51 @@
...
+
+            <androidx.appcompat.widget.AppCompatTextView
+                android:layout_width="0dp"
+                android:layout_height="match_parent"
+                android:layout_weight="1"
+                android:gravity="center_vertical"
+                android:paddingStart="@dimen/dimen_item_category_name_padding_start"
+                android:paddingEnd="0dp"
```
[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/res/layout/item_category.xml:40]: Снимок экрана 2022-01-24 в 01 20 00

[the remark this answers is not in the document]
Comment (VVMaxVV) [reply app/src/main/res/layout/item_category.xml:40]: if i add this line the warning disappears

**Review by darthoo — APPROVED** [review]

**Pull request merged**